etl/src/main/java/org/pih/hivmigration/etl/sql/ProgramMigrator.groovy

```diff
@@ -0,0 +1,252 @@
+        ''')
+
+        executeMysql("Add enrollment_date and outcome", '''
+            UPDATE hivmigration_programs_raw
+            SET
+                # Enrollment Date in program is earlier of minimum encounter and art start date
+                enrollment_date = LEAST (IFNULL(min_visit_date, art_start_date), IFNULL(art_start_date, min_visit_date)),                
+                outcome = CASE treatment_status
+                    WHEN 'died' THEN 'DIED'
+                    WHEN 'lost' THEN 'LOST' 
+                    WHEN 'abandoned' THEN 'LOST'
+                    WHEN 'transferred_out' THEN 'TRANSFERRED_OUT'
+                    WHEN 'treatment_refused' THEN 'TREATMENT_STOPPED'  # TODO: this might be its own outcome
+                    WHEN 'treatment_stopped' THEN 'TREATMENT_STOPPED'
+                    WHEN 'treatment_stopped_side_effects' THEN 'TREATMENT_STOPPED'
+                    WHEN 'treatment_stopped_other' THEN 'TREATMENT_STOPPED'
+                    END
+            ;
+        ''')
+
+
+        executeMysql("Add outcome date if outcome present", '''
+            UPDATE hivmigration_programs_raw
+            SET
+                # Outcome date is the most recent of treatment_status_date, regimen_outcome_date, and enrollment date
+                outcome_date = IFNULL(
+                    GREATEST (
+                        COALESCE(treatment_status_date, regimen_outcome_date, '1800-01-01'),
+                        IFNULL(regimen_outcome_date, '1800-01-01'),
+                        enrollment_date
+                    ),
+                    DATE_ADD(max_visit_date, INTERVAL 6 MONTH)
+                )
+            WHERE OUTCOME IS NOT NULL
+            ;
+        ''')
+
+        executeMysql("Fix ART start date in the 4-5 cases where it starts after the outcome", '''
+            UPDATE hivmigration_programs_raw
+            SET art_start_date = LEAST (art_start_date, outcome_date)
+            WHERE art_start_date IS NOT NULL AND outcome_date IS NOT NULL
+        ''')
+
+        executeMysql("Remove bad health center entries", '''
+            UPDATE hivmigration_programs_raw
+            SET health_center = IF (health_center = 0 OR health_center = 42, NULL, health_center),
+                starting_health_center = IF (starting_health_center = 0 OR starting_health_center = 42, NULL, starting_health_center)
+            ;
+        ''')
+
+        executeMysql("Ensure health_center is set and drop superfluous values for starting_health_center", '''
+            UPDATE hivmigration_programs_raw
+            SET
+                health_center = COALESCE(health_center, starting_health_center, 1),
+                starting_health_center = IF (
+                        starting_health_center = health_center OR
+                        health_center_transfer_date IS NULL OR
+                        health_center_transfer_date > outcome_date OR
+                        health_center_transfer_date < enrollment_date,
```

Comment:
Curious about the last two lines in the this clause (> outcome_date or < enrollment_date)... seems like, technically, bad data that might need to be cleaned up, but I could definitely see this being low value data we don't need to spend much time on.

Reply:
Yeah it seems like bad data is here being filtered out quite liberally. Agree that it would be good to review at some point.